### PR TITLE
Introduce static assert macro

### PR DESCRIPTION
**Pull request by kalj**
As identified in #3857, we often need to perform a `static_assert` while at the same time guarding it with checks of some conditions. This introduces a new macro `AssertStatic` similar to the existing ones, which simply wraps `static_assert`, while only enabling it under the right circumstances.

Questions:
1. Is the name good?
2. Is the documentation sufficient?

### CODE REVIEW

**Comment by drwells** [comment]
I am somewhat opposed to adding this since, in the near future, we will require C++11 and can just use the keyword.

IIRC the only reason we added the Microsoft guards was that MSVC got confused by the codimension cases and spat out spurious warnings, not that MSVC does not implement `static_assert`. We might be too pessimistic by always turning off static asserts for MSVC.

**Comment by tjhei** [comment]
> IIRC the only reason we added the Microsoft guards was that MSVC got confused by the codimension cases and spat out spurious warnings, not that MSVC does not implement static_assert.

Correct. It was spurious *errors* though: it would trigger this specific ``static_assert`` while trying different overloads for a function call.

File: include/deal.II/base/exceptions.h [hunk 1]
```diff
@@ -1228,6 +1228,23 @@ namespace StandardExceptions
                                            dealii::ExcCudaError(cudaGetErrorString(error_code)))
 #endif
 
+#if defined(DEAL_II_WITH_CXX11) || !defined(DEAL_II_MSVC)
```
Comment (bangerth) [inline include/deal.II/base/exceptions.h]: I think this should only test for C++11. In those (few) places where MSVC doesn't work, we should leave the `#ifndef MSVC` in place, but in all other places we can at least get rid of the check for C++11.

File: include/deal.II/base/exceptions.h [hunk 2]
```diff
@@ -1228,6 +1228,23 @@ namespace StandardExceptions
...
+ * than a wrapper around the C++11 <code>static_assert</code> declaration, which
+ * we need to disable for non-C++11 and for Visual Studio.
+ *
+ * Just like <code>static_assert</code>, it takes two arguments, a
+ * <code>constexpr</code> condition and an error message.
+ *
+ * @ingroup Exceptions
+ * @author Karl Ljungkvist, 2017
+ */
+#define AssertStatic(...) static_assert(__VA_ARGS__)
```
Comment (bangerth) [inline include/deal.II/base/exceptions.h]: The variadic macro is probably fine -- it requires C++11, but we know that we have it here.

The question is more whether you expect anyone to have more than one, comma separated expression. Can that happen? In other words, can `static_cast` take more than one comma separated expression?

Comment (bangerth) [inline include/deal.II/base/exceptions.h]: Oh, I see -- it is used as
```
  static_assert (condition, "error text");
```
So I guess this makes sense.

Comment (kalj) [inline include/deal.II/base/exceptions.h]: The problem is with expressions like `AssertStatic(mystruct<T,S>::value,"errormsg")`. The preprocessor doesn't handle the extra comma gracefully.

File: include/deal.II/base/exceptions.h (hunk 1 above)
Comment (kalj) [inline include/deal.II/base/exceptions.h]: Agreed.

**Comment by kalj** [comment]
@bangerth Does that mean that it should be "hidden", i.e. without documentation?

File: include/deal.II/base/exceptions.h (hunk 2 above)
Comment (bangerth) [inline include/deal.II/base/exceptions.h]: Ah, yes, good point.

**Comment by bangerth** [comment]
Good question. I would vote to indeed not advertise this macro via documentation, but I'm not sure. What do others think?

**Comment by drwells** [comment]
I support doing this with either documentation specifying that this is internal or just not adding any doxygen documentation for it.

It might not hurt to have this turn into `AssertThrow` when C++11 is not available: the compiler should be able to throw out unreachable branches anyway (unless, of course, someone uses it).

**Comment by bangerth** [comment]
No, I think converting it to `AssertThrow` is the wrong direction. First, `AssertThrow` can only appear in places where *expressions* are allowed, whereas `static_cast` is allowed in all places where *declarations* are allowed. Second, `static_cast` is a *compile time* assertion; turning them into `AssertThrow` statements would delay the error message to an unnecessarily late time.

**Comment by kalj** [comment]
@drwells @bangerth, Yes, and maybe the fact that it does not exactly behave as the other Assert* macros motivates even more to name it differently.

I'm updating this shorty with the all-caps name, and moving the MSVC-specific stuff back to the invocations.

**Comment by drwells** [comment]
Yes, this sounds like the right way to do things. :+1:

@bangerth You are right. Converting to `AssertThrow` wouldn't work anyway since we have the variadic thing, I think.

**Comment by kalj** [comment]
I think this is ready now. I just squashed and rebased it.

**Comment by bangerth** [comment]
Yes, nice, thanks! OK to merge once the tester is through.

**Pull request merged**